[XLA:GPU][ROCm] Refactor nvptx_backend_lib to support both NVPTX and AMDGPU

whchung opened the pull request:
Notice `nvptx_backend_lib` shall better be renamed as `gpu_backend_lib` but it is
skipped in this commit so minimize potential impacts to other XLA clients.

Subsequent PR would rename `nvptx_backend_lib` to `gpu_backend_lib`, and introduce complete AMDGPU-specific LLVM invocation logic.

- Determine platform-specific behaviors when constructing LLVM TargetMachine.

- Break CompileModuleToPtx into 3 functions:
  - NVPTXGetTargetMachine & GetTargetMachine : setup LLVM TargetMachine
  - LinkAndOptimizeModule : NVPTX-specific logic to link with libdevice, and run LLVM optimization passes.
  - EmitModuleToPTX : NVPTX-specific logic to drive LLVM NVPTX backend.

- Modify LinkLibdeviceIfNecessary to use LinkWithBitcodeVector.
  - LinkWithBitcodeVector would link a vector of paths to LLVM bitcode libs,
    this utility routine could support both NVPTX (libdevice) and AMDGPU
    (ROCm-Device-Libs).

whchung committed:
Refactor nvptx_backend_lib to support both NVPTX and AMDGPU

Notice nvptx_backend_lib shall better be renamed as gpu_backend_lib but it is
skipped in this commit so minimize potential impacts to other XLA clients.

- Created xla::gpu::nvptx namespace to store NVPTX-specific logic and values.

- Created xla::gpu::amdgpu namespace to store AMDGPU-specific logic and values.

- Extract platform-neutral logic to anonymous namespace.

- Pass StreamExecutor* from nvptx_compiler to nvptx_backend_lib to help
  determine platform-specific behaviors constructing LLVM TargetMachine.

- Break CompileModuleToPtx into 2 functions:
  - ConstructLLVMTargetMachineForModule : setup LLVM TargetMachine based on
      StreamExecutor* passed in from frontend.
  - nvptx::EmitModuleToPTX : NVPTX-specific logic to drive LLVM NVPTX backend.

- Modify LinkLibdeviceIfNecessary to use LinkWithBitcodeVector.
  - LinkWithBitcodeVector would link a vector of paths to LLVM bitcode libs,
    this utility routine could support both NVPTX (libdevice) and AMDGPU
    (ROCm-Device-Libs).

whchung commented:
The list of failures in `Linux GPU` should have no relationship with the PR.

whchung committed:
Fix ROCDL path processing logic

whchung committed:
Remove undesirable StreamExecutor from LLVM backend interface

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -202,29 +198,6 @@ void EmitBitcodeToFile(const Module& module, absl::string_view filename) {
 
-// Emits the given module to PTX. target_machine is an initialized TargetMachine
-// for the NVPTX target.
-string EmitModuleToPTX(Module* module, llvm::TargetMachine* target_machine) {
```
Please do not move functions around, because it makes the diffs harder to read.

whchung replied:
For `EmitModuleToPTX` the reason is that surrounding functions `AddOptimizationPasses`, `EmitBitcodeToFile`, and `FeedLLVMWithFlags` are generic enough be put into the anonymous namespace. But `EmitModuleToPTX` itself is NVPTX-specific so shall be better placed inside nvptx namespace. To avoid having namespaces everywhere I chose to move the function. I can put it back and wrap the function with nvptx namespace if you feel it be a better idea.

thomasjoerg replied:
I wonder whether namespaces are the best way to organize the code. Note that functions you take out of the anonymous namespace are visible outside this file. However, those functions are not a public API. Having a clear separation of NVPTX and AMDGPU-specific functions is definitely helpful for the reader. We could just use a code comment to make this clear without increasing visibility.

whchung replied:
@thomasjoerg In the latest round I've started adding `NVPTX` `AMDGPU` prefix to some platform-specific function. Please take a look at it and see if we should move into this direction.

thomasjoerg replied:
Yes, I think that's a better way of separating NVPTX vs AMDGPU-specific functions while restricting their visibility to this file. Let's go down that path and keep EmitModuleToPTX in the anonymous namespace. PTX in the name is enough to know this is for NVidia GPU. Same for other functions below.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -254,68 +227,70 @@ bool CouldNeedLibdevice(const llvm::Module& module) {
-    return xla::InternalError("libdevice not found at %s", libdevice_path);
-  }
+// Links the module with a vector of path to bitcode modules
+// The paths are guaranteed to exist.
```
The paths are passed by the caller. Hence, would it make sense to say that the caller must guarantee that the passes exist?

whchung replied:
thanks. will revise the comment.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -254,68 +227,70 @@ bool CouldNeedLibdevice(const llvm::Module& module) {
-        << libdevice_path;
-    return xla::InternalError("libdevice not found at %s", libdevice_path);
-  }
+// Links the module with a vector of path to bitcode modules
```
Nona-nit: End the sentence with a period. Also elsewhere.

whchung replied:
will do.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -254,68 +227,70 @@ bool CouldNeedLibdevice(const llvm::Module& module) {
+  llvm::Triple target_triple = llvm::Triple(module->getTargetTriple());
+  bool isNVPTX = target_triple.isNVPTX();
+
+  if (isNVPTX) {
```
if (isNVPTX) {
  ...
} else {
  ...
}
Is a highly repetitive pattern in this function. Please consider to fork the it into a NVPTX and AMD specific versions to improve readability and reduce the level of nesting.

whchung replied:
@thomasjoerg I think this is a design guideline I'd like to seek your opinion, as this request would also impact the frontend (`nvptx_compiler.cc`).

In the downstream fork I've been working on, originally we kept NVPTX and AMDGPU as separate modules, and they are enabled/disabled at configure-time via bazel `--config=` option. There were quite a few duplicated codes and @jlebar recommended unifying the logic as much as possible, hence I created this PR as a first step to make this `nvptx_backend_lib` to support both NVPTX and AMDGPU. And therefore I introduced this `isNVPTX` runtime check inside the function.

Based on your comments, would you think it'd be a good idea if I:

a) rename this function from `ConstructLLVMTargetMachine` to something like `LinkAndOptimizeModule`.
b) have two versions of `LinkAndOptimizeModule`, one in nvptx namespace, one in amdgpu namespace. And there would be some code duplication between the two versions, while keeping platform-specific logic separately.

whchung replied:
@thomasjoerg I've submitted a new version to address my points above. Please kindly have another round of review and see if you are okay with it. Thanks!

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -329,16 +304,30 @@ StatusOr<string> CompileModuleToPtx(llvm::Module* module,
+    if (isNVPTX) {
+      target_triple = llvm::Triple("nvptx64-unknown-unknown");
+    } else {
+      target_triple = llvm::Triple("amdgcn--amdhsa-amdgiz");
```
Are two dashes in "amdgcn--amdhsa-amdgiz" intended?

whchung replied:
Yes it's intended.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -254,68 +227,70 @@ bool CouldNeedLibdevice(const llvm::Module& module) {
-StatusOr<string> CompileModuleToPtx(llvm::Module* module,
-                                    std::pair<int, int> compute_capability,
+StatusOr<std::unique_ptr<llvm::TargetMachine>>
+ConstructLLVMTargetMachineForModule(llvm::Module* module,
```
The function name ConstructLLVMTargetMachineForModule seems misleading and sound similar to GetTargetMachine. The most important thing this function does is running assembling and running compiler passes on the module which is an output param. This needs to be reflected in the name. The target_machine is more of a by-product. I would be cleaner to pass target_machine in as param.

whchung replied:
How do you think if I rename this function from `ConstructLLVMTargetMachine` to something like `LinkAndOptimizeModule`?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -444,24 +510,80 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
 
 StatusOr<string> CompileToPtx(llvm::Module* module,
-                              std::pair<int, int> compute_capability,
+                              GpuVersion gpu_version,
```
Functions with "Ptx" in the name are NVidia-specific. Please move it under the nvptx namespace and keep using compute_capability rather than the generic GpuVersion.

whchung replied:
will do. In such case I'll also put AMDGPU-specific `CompileToHsaco` under `amdgpu` namespace.

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -63,39 +63,33 @@ limitations under the License.
+namespace amdgpu {
+
+// Inline threshold value to use in LLVM AMDGPU backend.
+const int kAMDGPUInlineThreshold = 1048576;
```
will do

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -243,7 +216,7 @@ void FeedLLVMWithFlags(const std::vector<string>& cl_opts) {
 // Returns whether the module could use any libdevice functions. This function
```
will do

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.h:
```diff
@@ -29,15 +30,16 @@ limitations under the License.
 namespace xla {
 namespace gpu {
 
+using GpuVersion = absl::variant<std::pair<int, int>, int>;
```
will do

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.h:
```diff
@@ -29,15 +30,16 @@ limitations under the License.
 
+using GpuVersion = absl::variant<std::pair<int, int>, int>;
+
 // Compiles the argument module and returns it. libdevice_dir_path is the parent
```
I intend to retain the interface of `CompileToPtx` and keep its comments as-is.

After this PR gets reviewed / revised and merged, I'll submit another PR for AMDGPU-specific logic. On AMDGPU side, a new interface `CompileToHsaco` would be introduced, and it would return `StatusOr<std::vector<uint8>>` as LLVM AMDGPU produces AMD GCN ISA binary in "HSA Code Object (HSACO)" format.

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -63,39 +63,33 @@ limitations under the License.
 namespace gpu {
-namespace {
+
+// Forward declaration for logic specific to LLVM NVPTX backend
```
I believe I met some errors keeping codes in the original spots. I'll try put them back and log what errors I met and post here.

whchung replied:
After re-arranging codes we no longer need forward declaration.

whchung committed:
Address code review comments.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -274,45 +329,26 @@ Status LinkLibdeviceIfNecessary(llvm::Module* module,
-  // Set the flush-denormals-to-zero flag on the module so the NVVM reflect pass
-  // can access it.
+  // Set the flush-denormals-to-zero flag on the module so the NVVM reflect
+  // pass can access it.
```
Nit: Please revert this change. Same below.

whchung replied:
will do

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -444,27 +482,169 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+                               GetROCDLPaths(amdgpu_version, rocdl_dir_path));
+}
+
+StatusOr<std::unique_ptr<llvm::TargetMachine>> LinkAndOptimizeModule(
```
In the previous round I asked you to remove the repetitive
    if (isNVPTX) { ... } else { ... }
pattern, because it makes the code hard to read. You are right when you say that this introduced code duplication.

However, I think we can get the best of both worlds as follows: Let's extract target-specific parts of this function, i.e. linking device libs and generating the target machine. What remains will be a target-independent LinkAndOptimizeModule function that can be called with the right inputs form CompileToPtx and the AMDGPU equivalent. Does that make sense?

whchung replied:
@thomasjoerg Ok, let me take a stab try extracting common logic from the 2 and pass in target-specific logic as callable functions.

On NVPTX side, target-specific logic are:
- link with libdevice if necessary
- add module flag
- add function attribute
- set target triple
- get target machine
- add optimization passes and set inliner threshold

On AMDGPU side, target-specific logic are:
- link with ROCm-Device-Libs if necessary
- set target triple
- get target machine
- add optimization passes and set inliner threshold

whchung replied:
@thomasjoerg I've submitted a new version. Please kindly review it again. Thanks!

whchung committed:
Address code review comments.

whchung committed:
Re-introduce GpuVersion

whchung committed:
Extract common logic from LinkAndOptimizeModule.

whchung committed:
Fix build errors.

whchung commented:
@thomasjoerg a gentle ping

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -444,27 +511,100 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
     TF_ASSIGN_OR_RETURN(
-        ptx, CompileModuleToPtx(module, compute_capability, hlo_module_config,
-                                libdevice_dir_path));
+        target_machine,
```
Let's call NVPTXGetTargetMachine here explicitly rather than plumbing the data through LinkAndOptimizeModule. This will read better and simplify LinkAndOptimizeModule's function signature.

whchung replied:
will do.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -319,6 +354,36 @@ StatusOr<string> CompileModuleToPtx(llvm::Module* module,
+  // Figure out the exact name of the processor as known to the NVPTX backend
+  // from the gpu_architecture flag.
+  return GetTargetMachine(
+      target_triple, GetSmName(absl::get<std::pair<int, int>>(gpu_version)),
```
Technically an AMDGPU GpuVersion could have been passed by the caller. Please handle this error case. Same for the AMDGPU-specific function. Alternatively, do not use the GpuVersion abstraction, so this error case can never happen.

whchung replied:
let me change `NVPTXGetTargetMachine` to take `std::pair<int, int>` and handle the case in `CompileToPtx`.

whchung committed:
Address code review comments.

- Move all utility functions into anonymous namespace.
- Refactor signature of LinkAndOptimizeModule. LLVMGetTargetMachine is invoked
  outside of it.
- Add checks for GpuVersion.

whchung commented:
@thomasjoerg submitted a new commit per your review comments. could you help re-review and approve again? Thanks a lot.

whchung commented:
@rthadur for some reason I couldn't see errors in `Linux GPU` target. Could you help shed some lights on the issue we are seeing?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,112 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+static std::vector<string> GetROCDLPaths(int amdgpu_version,
+                                         const string& rocdl_dir_path) {
+  // AMDGPU version-neutral bitcodes.
+  std::vector<string> rocdl_filename_vector{
```
Please initialize statically like this.
static rocdl_filenames* m = new std::vector<std::string>({...});

whchung replied:
done.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -240,25 +247,55 @@ void FeedLLVMWithFlags(const std::vector<string>& cl_opts) {
+
+  for (auto& bitcode_path : bitcode_path_vector) {
+    if (!tensorflow::Env::Default()->FileExists(bitcode_path).ok()) {
+      LOG(WARNING) << "bitcode module is required by this HLO module but was "
```
Please LOG an ERROR since this is considered and xla::InternalError.

whchung replied:
fixed.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -274,38 +311,18 @@ Status LinkLibdeviceIfNecessary(llvm::Module* module,
-  }
-  return Status::OK();
+  std::vector<string> libdevice_path_vector{libdevice_path};
+  return LinkWithBitcodeVector(module, libdevice_path_vector);
```
Would this work?
`return LinkWithBitcodeVector(module, {libdevice_path_vector});`

whchung replied:
done.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -274,38 +311,18 @@ Status LinkLibdeviceIfNecessary(llvm::Module* module,
-  TF_RETURN_IF_ERROR(
-      LinkLibdeviceIfNecessary(module, compute_capability, libdevice_dir_path));
+  TF_RETURN_IF_ERROR(LinkLibdeviceIfNecessary(
+      module, absl::get<std::pair<int, int>>(gpu_version),
```
Please handle the case were an AMDGPU-specific gpu_version is passed in, or revert to compute_capability pair.

whchung replied:
also applied `absl::get_if`

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,112 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+
+      // Lower optimize LLVM module to PTX.
+      TF_ASSIGN_OR_RETURN(ptx, EmitModuleToPTX(module, target_machine.get()));
+    }
```
Please return an error if compute_capability was not specified or do not use the generic GpuVersion.

whchung replied:
return `xla::InternalError` to handle the error.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,112 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+          NVPTXTargetModuleLinker, target_triple,
+          target_machine.get(), kDefaultInlineThreshold));
+
+      // Lower optimize LLVM module to PTX.
```
Type: optimized

whchung replied:
fixed.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,112 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+
+    auto compute_capability = absl::get_if<std::pair<int, int>>(&gpu_version);
+    if (compute_capability) {
+      llvm::Triple target_triple("nvptx64-unknown-unknown");
```
Please call it "default_target_triple" to clarify what this is used for.

whchung replied:
revised wording.

whchung committed:
Address code review comments.

whchung commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -204,7 +210,8 @@ void EmitBitcodeToFile(const Module& module, absl::string_view filename) {
 // Emits the given module to PTX. target_machine is an initialized TargetMachine
 // for the NVPTX target.
-string EmitModuleToPTX(Module* module, llvm::TargetMachine* target_machine) {
+StatusOr<string> EmitModuleToPTX(Module* module,
```
will do.

whchung commented:
@thomasjoerg Many thanks for your review. I just submitted a revised version.

whchung commented:
@thomasjoerg since PR #30238 has been merged I've rebased this PR. Now it looks more concise. Could you help re-review again? Thanks.

whchung committed:
Fix bazel dependencies.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,115 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+  // Add AMDGPU version-specific bitcodes.
+  result.push_back(tensorflow::io::JoinPath(
+      rocdl_dir_path, tensorflow::strings::StrCat(
+                          "oclc_isa_version_", amdgpu_version, ".amdgcn.bc")));
```
Please prefer absl over of tensorflow functions/types and use absl::StrCat here.

whchung replied:
fixed.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,115 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+Status LinkROCDLIfNecessary(llvm::Module* module, int amdgpu_version,
+                            const string& rocdl_dir_path) {
+  if (!CouldNeedDeviceBitcode(*module)) {
+    return tensorflow::Status::OK();
```
Please prefer Status::OK() (i.e. drop the tensorflow::) within the XLA codebase.

whchung replied:
fixed.

thomasjoerg reviewed (dismissed):
Looking good. Just two nits.

whchung committed:
Address code review comments.

whchung commented:
Failures in the 6 failing targets don't suggest anything to do with proposed changes introduced in this PR.

whchung commented:
@thomasjoerg would you mind re-approve this PR again? Thanks.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,115 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+  result.push_back(tensorflow::io::JoinPath(
+      rocdl_dir_path,
+      absl::StrCat("oclc_isa_version_", amdgpu_version, ".amdgcn.bc")));
+  return std::move(result);
```
Please drop the std::move.
Clang compiles fail with
error: moving a local object in a return statement prevents copy elision [-Werror,-Wpessimizing-move]

whchung replied:
thanks it's done.

whchung committed:
Address reviewer comments.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc:
```diff
@@ -446,25 +484,115 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
+std::unique_ptr<llvm::TargetMachine> AMDGPUGetTargetMachine(
+    llvm::Triple target_triple, int amdgpu_version,
+    const HloModuleConfig& hlo_module_config) {
+  return std::move(GetTargetMachine(target_triple,
```
Please remove the std::move to allow copy elision here too.

whchung replied:
it's done.

whchung committed:
Address reviewer comments.
